app/frontend/src/components/designer/FormDesigner.vue

```diff
@@ -369,15 +443,135 @@ export default {
     // ---------------------------------------------------------------------------------------------------
     init() {
       // Since change is triggered during loading
-      this.setDirtyFlag(false);
     },
-    onChangeMethod() {
+    onChangeMethod(_changed, _flags, _modified) {
       // Don't call an unnecessary action if already dirty
       if (!this.isDirty) this.setDirtyFlag(true);
+
+      // If the form changed but was not done so through the undo
+      // or redo button
+      if (!this.patchHistoryUndo && !this.patchHistoryRedo) {
+        // flags and modified are defined when a component is added
+        if (_flags !== undefined && _modified !== undefined) {
+          if (this.componentAdded !== null) {
+            this.addPatchToHistory();
+          } else {
+            this.resetHistoryFlags();
+          }
+        } else {
+          if (this.componentRemoved) {
+            // Component was removed..
+            this.addPatchToHistory();
+          } else if (this.componentMoved) {
+            this.addPatchToHistory();
+          } else {
+            this.resetHistoryFlags();
+          }
+        }
```

Comment:
- Double check the new variables such as componentAdded - do we need to be initalizing it as null, or can we keep it as just a simpler boolean? If we can do the latter, it can simplify the logic needed.
- This entire section of logic seems like it doesn't need to be branched because in most situations, you're either just calling `addPatchToHistory` or `resetHistoryFlags` methods when something changes (or not). You could potentially simplify this entire block to the following, ignoring what kind of change it was, and ignoring the `_flags` and `_modified` too (pseudocode... don't know if there's other side effects to worry about?):

```js
if (this.componentAdded || this.componentMoved || this.componentRemoved) {
  this.addPatchToHistory();
} else {
  this.resetHistoryFlags();
}
```

Reply:
first off, sorry for the wall of text, but this is for anyone else that may run into a similar issue!

the first one was a mistake on my part, initially i was storing the component in the componentAdded variable so that i could store the new component in the history after the user clicks "save". i've reverted it back to a boolean and now it is much cleaner, simpler, and unpacks/repacks correctly since we're using fast-json-patch.

the reason why it isn't simplified is because i wasn't sure at first how complex it would become. formio calls the change method for absolutely every change even if the user doesn't click save or if they click cancel/remove. we unfortunately need the `flags` and `modified` parameters to verify that the user clicked save and actually added the component to the form. i also had to create my own flags for when we started an action such as add, remove, move, and update. those flags are set to true when the actions begin in their respective functions `onAddComponentMethod`, `onRemoveComponentMethod`, `onUpdateComponentMethod` and then set to false once an action is completed.

for example, the save process alone looks like:
drag and drop -> `onUpdateComponentMethod()` -> `onChangeMethod(changed, undefined, undefined)` -> `onAddComponent()`-> `onUpdateComponentMethod()` -> `onChangeMethod(changed, undefined, undefined)`
save -> `onChangeMethod(changed, flags, modified)` -> `onChangeMethod(changed, undefined, undefined)`
canceled after drag and drop -> `onRemoveComponentMethod()` -> `onChangeMethod(changed, undefined, undefined)`
removed after drag and drop -> `onRemoveComponentMethod()` -> `onChangeMethod(changed, undefined, undefined)`

after writing all this, i realized we don't really need fine-grained control over the actions since we're not writing the json comparator/patcher. i've removed the `onUpdateComponentMethod` method since an update is similar to an "add" with the `flags` and `modified` parameters defined. if we did want fine-grained control, then we could obtain the information for exactly what is being added/updated/removed and know if the action was saved or canceled/removed.